lib/libvfio-user.c

```diff
@@ -920,133 +894,284 @@ static int
+static uint32_t
+device_feature_flags_supported(vfu_ctx_t *vfu_ctx, uint32_t feature)
 {
-    struct vfio_user_dirty_pages *dirty_pages_in;
-    struct vfio_user_dirty_pages *dirty_pages_out;
-    struct vfio_user_bitmap_range *range_in;
-    struct vfio_user_bitmap_range *range_out;
-    size_t argsz;
-    int ret;
+    if (vfu_ctx->migration == NULL) {
+        /*
+         * All of the current features require migration.
+         */
+        return 0;
+    }
 
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_MIGRATION:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT:
+            return VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE:
+            return VFIO_DEVICE_FEATURE_GET
+                | VFIO_DEVICE_FEATURE_SET
+                | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_START:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP:
+            return VFIO_DEVICE_FEATURE_SET | VFIO_DEVICE_FEATURE_PROBE;
+        default:
+            return 0;
+    };
+}
 
-    dirty_pages_in = msg->in.iov.iov_base;
+static bool
+is_migration_feature(uint32_t feature)
+{
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_MIGRATION:
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE:
+            return true;
+    }
 
-    if (msg->in.iov.iov_len < sizeof(*dirty_pages_in) + sizeof(*range_in) ||
-        dirty_pages_in->argsz > SERVER_MAX_DATA_XFER_SIZE ||
-        dirty_pages_in->argsz < sizeof(*dirty_pages_out)) {
-        vfu_log(vfu_ctx, LOG_ERR, "invalid message size=%zu argsz=%u",
-                msg->in.iov.iov_len, dirty_pages_in->argsz);
-        return ERROR_INT(EINVAL);
+    return false;
+}
+
+static bool
+is_dma_feature(uint32_t feature)
+{
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_START:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT:
+            return true;
     }
 
-    range_in = msg->in.iov.iov_base + sizeof(*dirty_pages_in);
+    return false;
+}
 
-    /*
-     * range_in is client-controlled, but we only need to protect against
-     * overflow here: we'll take MIN() against a validated value next, and
-     * dma_controller_dirty_page_get() will validate the actual ->bitmap.size
-     * value later, anyway.
-     */
-    argsz = satadd_u64(sizeof(*dirty_pages_out) + sizeof(*range_out),
-                       range_in->bitmap.size);
+static int
+handle_migration_device_feature_get(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
+                                    struct vfio_user_device_feature *req)
+{
+    // all supported outgoing data is currently the same size as 
+    //   vfio_user_device_feature_migration
+    msg->out.iov.iov_len = sizeof(struct vfio_user_device_feature)
+        + sizeof(struct vfio_user_device_feature_migration);
+
+    if (req->argsz < msg->out.iov.iov_len) {
+        msg->out.iov.iov_len = 0;
```

Comment:
Here and everywhere else, can you avoid setting `msg->out.iov.iov_len` up until the last minute in order to avoid having to clear it?

I checked the code and it turns out returning -1 isn't enough, the transport code will still send data in case of error which I don't think we ever need to do. Could you raise an issue to fix that?

Reply:
I feel like `msg->out.iov.iov_len` should be set before allocating memory for `msg->out.iov.iov_base` so that it's never wrong...?

Reply:
agreed it should be set first.

We should probably explicitly allow it to still be set even if iov_base is NULL, when we free  msg.

Reply:
I don't know what you mean by that. I'm saying we should explicitly document that iov.len can be non-zero even if iov.iov_base is NULL, in free_msg() - a comment there is fine.
Then we don't need to reset iov_len in error cases.

Reply:
If I remember correctly I was resetting iov_len in error cases so that we don't send any data to the client (other than the header with the error flag and whatnot), because whatever data we may or may not have already allocated would be invalid. Is that incorrect behaviour? @jlevon

Reply:
we should be ignoring iov_len if it's set, but iov_base is NULL - that is, there is no data.

Reply:
> I don't know what you mean by that. 

I mean that it's already initialized so the user doesn't need to be explicitly set to 0 in the error case (assuming they haven't touched it).

Reply:
> I'm saying we should explicitly document that iov.len can be non-zero even if iov.iov_base is NULL, in free_msg() - a comment there is fine.
Then we don't need to reset iov_len in error cases.

That's a good idea.

Reply:
So we should be freeing and setting iov_base to NULL instead of setting iov_len to 0?

Reply:
look at the context. you haven't even set iov_base yet, so it's already NULL.

I'm saying make these specific changes:

 - remove this line setting iov_len to 0
 - same for other places that are error paths in e.g. handle_device_feature
 - if iov_base is NULL (and iovecs isn't set), clear out iov_len to 0 in do_reply(), along with a comment as to why...

Reply:
There are places where we set iov_len to 0 after allocating iov_base, e.g. at the end of handle_dma_device_feature_get and handle_mig_data_read... What do you suggest I do in these cases?

Reply:
Taking a step back, do we ever send a payload in case or error? If not, then `do_reply` could skip sending other iovecs. Or would that hide bugs?

Reply:
I'm not entirely sure - I think my preference is to remain explicit about removing the payload. Now I look at `do_reply` I don't think we can just have iov_base be NULL because it requires two iovecs, the second being a payload (which can be empty, as we do by setting iov_len to 0).

Reply:
@jlevon I've added that, can you take a look?

Reply:
#777